### ISSUE TITLE
Remove `ByteString` from `typing` and `collections.abc`

### ISSUE THREAD

**Issue by sobolevn** [python/cpython#118803]
# Feature or enhancement

It was deprecated and scheduled for removal in 3.14

Original issue: https://github.com/python/cpython/issues/91896
Ref: https://github.com/python/cpython/commit/09b7695f12f2b44d2df6898407d7e68dd9493ed4

### Linked PRs
* gh-118804
* gh-138990
* gh-138995
* gh-139115
* gh-139127
* gh-139136
* gh-139137
* gh-139171

**Comment by cdce8p** [comment]
Tbh I was a bit surprised that it didn't raise any DeprecationWarning on 3.13. Looking at #118804 it seems that was only emitted on `isinstance` checks. Most examples I've seen while testing 3.14 just import it at runtime to use as annotation.

It's a bit unfortunate to not have gotten a warning for it earlier as a few packages will now require immediate fixes for 3.14. Not sure if it would be worth it to revert the removal and add a proper DeprecationWarning on import (i.e. in `typing.__getattr__`)?

**Comment by cdce8p** [comment]
> What are some examples of libraries still using this?

I've only seen it a couple of times so far. Most cases have already been fixed for some time now:
- `crc` -> Fixed in `5.0` (https://github.com/Nicoretti/crc/commit/2ae4871a8f391ffbb267abfed78e38c430afba55). For Home Assistant we have a dependency which still (unnecessarily) pins the version to `<5`. Will work to update that instead.
- `pynacl` -> Fixed last year in https://github.com/pyca/pynacl/pull/848 but no new release yet.

**Comment by JelleZijlstra** [comment]
Anyone interested in sending a PR to restore it for 3.14?

**Comment by sobolevn** [comment]
I think that we should check top 10000 packages for its usages. If there are many existing ones - it might be a good idea to restore it. But, if there are just several - then we should probably move forward, it was deprecated for 5 versions already, there was a lot of time to change the code `.. deprecated-removed:: 3.9 3.14`. It is not hard to fix in the library code as `ByteString = bytes | bytearray`.

**Comment by jsbueno** [comment]
Searching around, I could not find the reasoning that led to this deprecation and removal altogether. Could someone point to a reference?  (discourse/mail group links accepted)

**Comment by hauntsaninja** [comment]
It gets a little into ugly weeds, but see https://github.com/python/cpython/issues/91896 and https://github.com/python/cpython/issues/102092

**Comment by hauntsaninja** [comment]
Like Alex and Jelle, I'm personally pretty happy to restore it for 3.14 (although it's getting late, so not sure what the process looks like)

**Comment by brianschubert** [comment]
Marking as a release blocker pending a decision about whether to restore it

**Comment by vivodi** [comment]
> Like Alex and Jelle, I'm personally pretty happy to restore it for 3.14 (although it's getting late, so not sure what the process looks like)

All the libraries I know of have already removed the use of `ByteString`, and it wasn’t difficult. I don’t think it’s necessary to bring it back.

**Comment by JelleZijlstra** [comment]
Yeah I think the ship has sailed, we're deep in the RC period now and there haven't been many more complaints. It wouldn't have hurt to wait another version or two, but at this point it might be more disruptive to put it back.

**Comment by jsbueno** [comment]
So, indeed  - looking there, it **really** looks like we got some escalation which may end up hurting people and code around, and removing some possibilities in the language.

I was looking for a rationally for its removal - and all I found is that it was a poorly done thing - but the concept is still important. 

As it is, @rhettinger initial comment there is 

> One other thought: the name ByteString wasn't very good to begin as is suggests bytes | str. If you really must do this, it would be better to create a new, well-named ABC with the requisite methods and even leave the old ABC around (it isn't hurting anything) or deprecate it.

So - he agrees there the concept is important as an ABC and typing - just the naming - and I dare adding, in Hettinger's opinion - was bad.  (I won't make a fuzz about defending the name, but it is pretty clear to me   = "bytestrings" in contrast to both "text string" and "bytes" as a sequence which has bytes as elements. )

Anyway, it should be upgraded to a _real_ abc which would allow other byte-yielding sequences to be registered (and include others by default, such as memoryview instances) - instead it is getting removed from the language altogether. 

(And I believe it is too late now to reverse it back in time for 3.14 - people are usually strict about changes during the RC cycle  - we might try get a `ByteSequence` proper ABC to Python 3.15 )

**Comment by hauntsaninja** [comment]
This removal is the kind of thing that would annoy me a lot as a user, e.g. all [708 (non-transitive) reverse deps](https://www.wheelodex.org/projects/pynacl/) of pynacl [will likely be unable](https://github.com/pyca/pynacl/issues/839) to use 3.14 on day 1

If you want to introduce some other ABC with better, different semantics, I'd recommend opening a new issue. We are eventually going to remove `ByteString`, the question is just when.

**Comment by vivodi** [comment]
> This removal is the kind of thing that would annoy me a lot as a user, e.g. all [708 (non-transitive) reverse deps](https://www.wheelodex.org/projects/pynacl/) of pynacl [will likely be unable](https://github.com/pyca/pynacl/issues/839) to use 3.14 on day 1
> 
> If you want to introduce some other ABC with better, different semantics, I'd recommend opening a new issue. We are eventually going to remove `ByteString`, the question is just when.

pynacl has supported Python 3.14 for a while, it just hasn’t been released yet. I believe a new version will be released before 3.14 comes out.

See https://github.com/pyca/pynacl/pull/848

**Comment by sobolevn** [comment]
Oh, it is pretty complex :)

Looks like we didn't communicate several things in a clear manner:
1. Why this was deprecated (since this question was raised here: https://github.com/python/cpython/issues/118803#issuecomment-3256653827)
2. That it was deprecated to be imported. It was only a warning when `ByteString` was initialized

Sorry about that!

The fix in https://github.com/pyca/pynacl/pull/848 seems pretty easy to do, though. I don't think that restoring it would make things easier and not harder :(

**Comment by hugovk** [comment]
> pynacl has supported Python 3.14 for a while, it just hasn’t been released yet. I believe a new version will be released before 3.14 comes out.

PyNaCl 1.6.0 has now been released: https://github.com/pyca/pynacl/issues/839#issuecomment-3276864012

Are there any other major libraries needing an update?

**Comment by hugovk** [comment]
Just talked with @AlexWaygood at the sprint, and it would be good to emit the proper deprecation warnings before removing this in a later release. Alex will do this before the 3.14 RC on Wednesday.

**Comment by AlexWaygood** [comment]
We've now done everything we plan to do here until 3.17
